File: tests/standalone/potint4/main.py
import numpy as np

from engines.charge.potint4b import potint4b

cases = [
    (
        "single_tri_single_obs",
        np.array([[0.0, 0.0, 0.0]]),
        np.array([[1.0, 0.0, 0.0]]),
        np.array([[0.0, 1.0, 0.0]]),
        np.array([[0.1, 0.1, 0.2]]),
    ),
    (
        "single_tri_multi_obs",
        np.array([[0.0, 0.0, 0.0]]),
        np.array([[1.0, 0.0, 0.0]]),
        np.array([[0.0, 1.0, 0.0]]),
        np.array([[0.1, 0.1, 0.2], [0.5, 0.5, 1.0], [2.0, 2.0, 2.0]]),
    ),
    (
        "multi_tri_single_obs",
        np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]]),
        np.array([[1.0, 0.0, 0.0], [1.0, 0.0, 1.0]]),
        np.array([[0.0, 1.0, 0.0], [0.0, 1.0, 1.0]]),
        np.array([[0.2, 0.2, 0.5]]),
    ),
    (
        "multi_tri_multi_obs",
        np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0], [1.0, 0.0, 0.0]]),
        np.array([[1.0, 0.0, 0.0], [1.0, 0.0, 1.0], [2.0, 0.0, 0.0]]),
        np.array([[0.0, 1.0, 0.0], [0.0, 1.0, 1.0], [1.0, 1.0, 0.0]]),
        np.array([[0.1, 0.1, 0.2], [0.9, 0.1, 0.2], [0.5, 0.5, 0.5]]),
    ),
    (
        "on_vertex",
        np.array([[0.0, 0.0, 0.0]]),
        np.array([[1.0, 0.0, 0.0]]),
        np.array([[0.0, 1.0, 0.0]]),
        np.array([[0.0, 0.0, 0.0]]),
    ),
    (
        "degenerate",
        np.array([[0.0, 0.0, 0.0]]),
        np.array([[1.0, 1.0, 1.0]]),
        np.array([[2.0, 2.0, 2.0]]),
        np.array([[1.0, 0.0, 0.0]]),
    ),
]

for name, r1, r2, r3, obs in cases:
    print("CASE:", name)
    print("r1.shape:", r1.shape, "r1:\n", r1)
    print("r2.shape:", r2.shape, "r2:\n", r2)
    print("r3.shape:", r3.shape, "r3:\n", r3)
    print("obsPoint.shape:", obs.shape, "obsPoint:\n", obs)
    try:
        out = potint4b(r1, r2, r3, obs)
        print("output.shape:", out.shape if isinstance(out, np.ndarray) else "N/A")
        print("output:\n", out)
        success = True
    except Exception as e:
        print("ERROR:", str(e))
        success = False
    print("success:", success)
    print("-" * 80)

File: tests/standalone/potint4/main.m
cases = {};

cases{end+1} = struct('name','single_tri_single_obs', ...
    'r1', [0 0 0], 'r2', [1 0 0], 'r3', [0 1 0], 'obsPoint', [0.1 0.1 0.2]);

cases{end+1} = struct('name','single_tri_multi_obs', ...
    'r1', [0 0 0], 'r2', [1 0 0], 'r3', [0 1 0], 'obsPoint', [0.1 0.1 0.2; 0.5 0.5 1.0; 2 2 2]);

cases{end+1} = struct('name','multi_tri_single_obs', ...
    'r1', [0 0 0; 0 0 1], 'r2', [1 0 0; 1 0 1], 'r3', [0 1 0; 0 1 1], 'obsPoint', [0.2 0.2 0.5]);

cases{end+1} = struct('name','multi_tri_multi_obs', ...
    'r1', [0 0 0; 0 0 1; 1 0 0], 'r2', [1 0 0;1 0 1;2 0 0], 'r3', [0 1 0;0 1 1;1 1 0], ...
    'obsPoint', [0.1 0.1 0.2; 0.9 0.1 0.2; 0.5 0.5 0.5]);

cases{end+1} = struct('name','on_vertex', ...
    'r1', [0 0 0], 'r2', [1 0 0], 'r3', [0 1 0], 'obsPoint', [0 0 0]);

cases{end+1} = struct('name','degenerate', ...
    'r1', [0 0 0], 'r2', [1 1 1], 'r3', [2 2 2], 'obsPoint', [1 0 0]);

for k = 1:numel(cases)
    c = cases{k};
    fprintf('CASE: %s\n', c.name);
    disp('r1:'); disp(c.r1); disp('r2:'); disp(c.r2); disp('r3:'); disp(c.r3);
    disp('obsPoint:'); disp(c.obsPoint);
    try
        out = potint4b(c.r1, c.r2, c.r3, c.obsPoint);
        disp('output:'); disp(out);
        success = 1;
    catch ME
        fprintf('ERROR: %s\n', ME.message);
        success = 0;
    end
    fprintf('success: %d\n', success);
    fprintf('%s\n', repmat('-',1,80));
end
